### PR TITLE
Refactor: group some fields into an abstracted type

### PR DESCRIPTION
**Pull request by vext01**
This is a refactoring that @ptersilie and I have superficially discussed in the past.

It adds a little more code than I expected, but I hope that the added safety aspects outweigh that cost?

See the first commit message for details.

### CODE REVIEW

**Comment by ptersilie** [comment]
Looks fine to me. But @ltratt might have some comments on the design?

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 1]
```diff
@@ -320,23 +368,23 @@ class JITModBuilder {
     Builder.SetInsertPoint(DstBB);
 
     // Iterate over the trace and stitch together all traced blocks.
-    for (size_t Idx = 0; Idx < Len; Idx++) {
-      auto FuncName = FuncNames[Idx];
-
-      if (ExpectUnmappable && (FuncName == nullptr)) {
+    for (size_t Idx = 0; Idx < InpTrace.Length(); Idx++) {
+      Optional<IRBlock> MaybeIB = InpTrace[Idx];
+      if (ExpectUnmappable && (!MaybeIB.hasValue())) {
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:373]: I don't think the inner brackets around the `!` are needed.

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 2]
```diff
@@ -24,12 +24,65 @@ void dumpValueAndExit(const char *Msg, Value *V) {
...
+// A function name and basic block index pair that identifies a block in the
+// AOT LLVM IR.
+struct IRBlock {
+  // A non-null pointer to the function name.
+  char *FuncName;
+  // The index of the block in the parent LLVM function.
+  size_t BBIdx;
+};
+
+// Describes the software or hardware trace to be compiled using LLVM.
```
Comment (ltratt) [inline ykllvmwrap/src/jitmodbuilder.cc:36]: "Describes" seems an odd term here: I'm actually not sure that the struct is accurately named? Judging by the class fields, it seems more like a "block to function" mapper or ... ?

Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: It really does express the collected trace. The pair `(FuncNames[Idx], BBs[Idx])` identifies the LLVM IR block at position `Idx` in the trace (as long as `FuncNames[Idx] != nullptr`).

 The choice of data structures is a little odd due to shoe-horning through the C ABI (whilst trying not to add a new struct type to both the C side and the rust side).

Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: We could add a `repr(C)` type to Rust and a mirror type to C++ if you think that's cleaner?

Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: > We could add a repr(C) type to Rust and a mirror type to C++ if you think that's cleaner?

I have no idea! Please do as you see fit here.

Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: What do you think @ptersilie

Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: There is perhaps a small performance advantage to using a repr(C) type, as we then don't have to materialise them from an array once in C++...

Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: Maybe that's something for a future PR?

Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: If you think it's a future PR thing, then let's file an issue for it and move on?

**Comment by vext01** [comment]
bfa1c25 addresses outstanding comments.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 2 above)
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: I guess https://github.com/ykjit/yk/issues/419 is the relevant issue?

**Comment by ltratt** [comment]
Please squash.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 2 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:36]: Yes. I added a comment with a link in bfa1c25.

**Comment by vext01** [comment]
splat.

**Pull request merged**